### DISCUSSION TITLE
Proposal: Single Function Classes

### DISCUSSION THREAD

**Discussion by repinvv** [dotnet/csharplang#6617] [Language Ideas]
In pursuit of "single responsibility principle", i do break the code down to small classes, which in some cases end up having a single method/function. Here is an example:

```
internal class FieldTypeService(SomeDependency dependency)
{
  public Type DetermineFieldType(string type, bool isNullable)
  {  
   ... // implement 
  }
}
```

So the class is almost a C-style function, but it has dependencies, which enables isolated testing, unlike actual C-function, which calls other functions freely and uncontrolled.
The nuisance here is that i have to come up with 2 names, one for class name, and one for method name.
Here is the suggestion i have for the syntax:

```
internal class DetermineFieldType(SomeDependency dependency)(string type, bool isNullable)
{
... // implement
}
```

Suggested syntax for using such a class as a dependency:

```
internal class MyClass(DetermineFieldType determineFieldType)
{
  void SomeMethod()
  {
    determineFieldType("someType", false);
  }
}
```

or, if i am to create an instance manually (which i never do, but anyway):

```
var output = new DetermineFieldType(new SomeDependency())("someType", false);
```

Compiler would transform a body of a class into a method named "Invoke" or something like that.

What i could not come up with yet is where to put readonly fields in this syntax. Maybe it could go without any fields at all, except those that come from primary constructor.

**Comment by paulomorgado** [comment]
Can you share more compelling examples for this feature? I can't see the value in it.

**Comment by repinvv** [comment]
The value is in eliminating double naming. 
class "NameGetter" with method "GetName" would become a single function class "GetName".  And having an instance of the class is almost like having a delegate, i would just execute it as "getName()" instead of "nameGetter.GetName()". 
So it would be a further improvement of "primary constructor" syntax, i.e. purpose of it is to further reduce the code verbosity.

**Comment by MgSam** [comment]
Why make something a class if it has no state to keep track of and only has a single method? I don't see the benefit.

**Reply by thomaseyde** [reply]
Where would you put it, then? Namespace functions are still not a thing in C#. 

Besides, the dependency is the state. And this is one way to do partial function application in C#:

1. With a class. More ceremony, easy to reason about. Easy to pass around.
2. With a delegate. Less ceremony, harder to reason about. Harder to pass around.

Both implementations are technically equally easy to pass around, but a `Func<int, int>` can be anything. It can be redefined anywhere in the call stack, so finding the implementation when you need it is hard. That is much easier with a class: Simply navigate to the class.

```c#
public class Partial_application
{
    static readonly Doubler Doubler = new(2);
    static readonly Func<int, int> Double = i => Calculator.Multiply(2, i);

    [Fact]
    public void With_class()
    {
        Assert.Equal(6, Calculator.Multiply(2, 3));

        Assert.Equal(4, Doubler.Double(2));
        Assert.Equal(16, Doubler.Double(8));
    }

    [Fact]
    public void With_lambda()
    {
        Assert.Equal(4, Double(2));
        Assert.Equal(8, Double(16));
    }
}

public record Doubler(int Value)
{
    public int Double(int i) => Calculator.Multiply(Value, i);
}

static class Calculator
{
    public static int Multiply(int a, int b) => a * b;
}
```

**Comment by mburbea** [comment]
This sounds like a case where you require an `IFoo` which is an `interface` with one method, when you really wanted to just ask for `delegate` of the single method. There is no silver bullet. While the delegate is lightweight and doesn't have a lot of baggage to carry around, the moment you want to add another method to the interface you now need to switch to a full interface.

**Comment by aluanhaddad** [comment]
If I understand your proposal correctly, your proposal, semantically and syntactically, is actually very close to partial function application invocation support (or curried function definition support). I do not think a sugar for classes is the way to go here.

Here is something I find quite useful from time to time (naming edited slightly based on your use case):

``` C#
public static Func<TDep, T1, T2, TResult> Apply(this Func<TDep, T1, T2, TResult> f, TDep dep) =>
    (t1, t2) => f(dep, t1, t2);
```

Then if you have the following method:

``` C#
public Type DetermineFieldType(SomeDependency dependency, string type, bool isNullable) { ... }
```

You can partially apply it, but the problem is a delegate is required and often must be explicitly declared to use this.

``` C#
Func<SomeDependency, string, bool, Type> f = DetermineFieldType;
var determineFieldType = f.Apply(new SomeDependency());
```

Your proposed syntax is reminiscent of curried function definitions in Scala.

``` Scala
def DetermineFieldType(SomeDependency dependency)(string type, bool isNullable): Type = { ... }
```

Anyway, I think it would be great to have better definition and consumption syntax for this pattern in C#.
This could be done by allowing extension methods to target method groups, perhaps by the implicit construction of a delegate (not sure how difficult this would be to implement), or adding support for multiple argument lists in function definitions.

**Comment by repinvv** [comment]
Right, it does look close to curried function, now that you mention it.
Btw, i am using such classes with IoC, i.e. dependency in that constructor is injected, which would not be possible with partially applied function example that you gave there.

Not sure if that falls into overall curried functions concept or if IoC/DI pattern is even used in Functional Languages, never did learn FL's myself.

**Comment by aluanhaddad** [comment]
Actually functional techniques often implement Dependency Injection / Inversion of Control in this manner. You are correct that this does not support constructor injection, I was thinking of its use with the factory method pattern. 
Consider one has something like this:

``` C#
interface IVehiclePart { ... }

interface IVehicle { ... }

class Car: IVehicle { 
    public Car(string model, string manufacturer, IEnumerable<IVehiclePart> parts) { ... } 
}

interface IVehicleComponents
{
    string Manufacturer { get; }
    IEnumerable<IVehiclePart> Parts { get; }
}

class VehicleComponents
{ 
    public VehicleComponents(string manufacturer, IEnumerable<IVehiclePart> parts) { ... } 
}

delegate TVehicle VehicleFactory<TVehicle>(string model) where TVehicle: IVehicle;

public VehicleFactory<TCar> CreateCarFactory<TCar>(IVehicleComponents components) 
    where TCar: IVehicle=>
    model => new Car(model, components.Manufacturer, components.Parts);

```

Then they could write:

``` C#
var hondaParts = new List<IVehiclePart> { ... };
var hondaFactory = CreateCarFactory(new VehicleComponents("Honda", hondaParts);

var hondas = new [] { hondaFactory("Civic"), hondaFactory("CRV") };
```

or:

``` C#
using static Enumerable;

var hondaParts = new List<IVehiclePart> { ... };
var hondaFactory = CreateCarFactory(new VehicleComponents("Honda", hondaParts);
var models = new[] { new { Name = "Civic", Quantity = 4 }, new { Name = "CRV", Quantity = 2 } };
var hondas =  models.SelectMany(model => Repeat(model.Name, Model.Quantity), hondaFactory);
```

**Comment by dsaf** [comment]
@repinvv 

Do you mostly follow just the _single responsibility principle_ or other SOLID principles as well? The reason I am asking is because from my personal experience I would expect you to reason about single-method interfaces rather than single-method classes. Based on your example - internal concrete implementations - a simple static internal class full of methods would do just fine (effectively emulating F# module).

I would agree though, that a single-method interface is a stronger contract and cannot be replaced by a function.

**Comment by iam3yal** [comment]
Related #95 

p.s. SRP doesn't really relate to having a single method on whatever type, you can have a single method on a class that would break this principle whereas multiple methods wouldn't break it.

**Comment by CyrusNajmabadi** [comment]
This appears to be a language suggestion.  Moving to discussion on csharplang.
